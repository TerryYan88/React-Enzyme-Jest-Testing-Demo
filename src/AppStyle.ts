import styled from "styled-components";

export const AppContainer = styled.div`
    .main{
        width:100%;
        max-width:980px;
        margin:50px auto 80px;
    }
`

