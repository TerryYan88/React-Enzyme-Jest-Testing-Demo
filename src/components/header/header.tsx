import React from 'react';
import { HeaderContainer } from "./headerStyle";

const Header = () => {
    return (
        <HeaderContainer data-test="headerComponent">
            <div className="wrap">
                <div className="logo">
                    {/* data-test is to tell other developers 
                    can not change logImg, 
                    beacuse it is in a test */}
                    <img data-test="logoImg" src="http://localhost:3000/assets/2.png" alt="Logo" />
                </div>
            </div>
        </HeaderContainer>
    )
}

export default Header;