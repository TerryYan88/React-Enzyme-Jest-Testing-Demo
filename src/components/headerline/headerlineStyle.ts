import styled from "styled-components";


export const HeaderLineContainer = styled.div`
    

`