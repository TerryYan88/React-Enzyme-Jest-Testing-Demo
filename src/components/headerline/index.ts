export {default as HeaderLine} from "./headerline";