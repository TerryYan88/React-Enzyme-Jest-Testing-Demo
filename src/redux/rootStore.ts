import {createStore,applyMiddleware} from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import createSagaMiddleware,{SagaMiddleware} from "redux-saga";
import {rootReducers} from "./reducers/rootReducers";
import { createLogger } from "redux-logger";
import {AppActions} from "./interface/actions-types";
import {rootSagas} from "./sagas/sagas";
const sagaMiddleware = createSagaMiddleware();
const logger = createLogger();
export type AppState = ReturnType<typeof rootReducers>;

export const store = createStore<AppState,AppActions,{},{}>(
    rootReducers,
    composeWithDevTools(applyMiddleware(sagaMiddleware as SagaMiddleware<AppState>,logger )));
sagaMiddleware.run(rootSagas)
