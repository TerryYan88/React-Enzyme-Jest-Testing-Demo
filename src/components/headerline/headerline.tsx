/* eslint-disable @typescript-eslint/no-useless-constructor */

import React, { Component } from "react";
import { HeaderLineContainer } from "./headerlineStyle";
import PropTypes from "prop-types";

interface ItempArr {
    fName: string,
    lName: string,
    email: string,
    age: number,
    onlineStatus: boolean,
}

interface HeaderLineProps {
    header: string;
    desc: string;
    tempArr: ItempArr[];
}


class HeaderLine extends Component<React.PropsWithChildren<HeaderLineProps>> {
    static propTypes = {
        header: PropTypes.string.isRequired,
        desc: PropTypes.string.isRequired,
        tempArr: PropTypes.arrayOf(PropTypes.shape({
            fName: PropTypes.string.isRequired,
            lName: PropTypes.string.isRequired,
            email: PropTypes.string.isRequired,
            age: PropTypes.number.isRequired,
            onlineStatus: PropTypes.bool.isRequired,
        })).isRequired,
    }
    constructor(props: HeaderLineProps) {
        super(props);
    }
    render() {
        const { header, desc } = this.props;

        return (
            <HeaderLineContainer data-test="HeadlineComponent">
                {
                    header && <h1 data-test="header">{header}</h1>
                }
                {
                    desc && <p data-test="desc">{desc}</p>
                }
            </HeaderLineContainer>
        )
    }
}

export default HeaderLine;