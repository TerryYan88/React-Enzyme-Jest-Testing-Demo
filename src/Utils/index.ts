/* eslint-disable react/forbid-foreign-prop-types */
import {ShallowWrapper} from "enzyme";
import checkPropTypes from "check-prop-types";
import { applyMiddleware,createStore } from "redux";
import { rootReducer } from "../redux/reducers/rootReducer";


export const testStore = (initialState={}) =>{
    const createStoreWithMiddleware = applyMiddleware()(createStore)
    return createStoreWithMiddleware(rootReducer,initialState);
}


export const findByTestAttr = (component: ShallowWrapper,attr:string)=>{
    const wrapper = component.find(`[data-test="${attr}"]`);
    return wrapper;
}

export const checkProps = (component:any,expectedProps:{})=>{
    const propsErr = checkPropTypes(component.propTypes,expectedProps,'props',component.name);
    return propsErr;
}


