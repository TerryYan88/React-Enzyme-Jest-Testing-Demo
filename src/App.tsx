import { memo } from "react";
import { useDispatch, useSelector } from "react-redux";
import {Header} from "./components/header";
import {HeaderLine} from "./components/headerline";
import {Button} from "./components/button";
import {AppContainer} from "./AppStyle";
import ListItem from "./components/list/List";
import { ListActionsTypes } from "./redux/interface/saga-type";
import { AppState } from "./redux/rootStore";


const tempArr = [
  {
    fName:"TT",
    lName:"YY",
    email:"1xxxxx@gmail.com",
    age:24,
    onlineStatus:true,
  }
]

function App() {
  const header = "Posts";
  const desc = "Click the button to render posts!";
  const buttonText = "Click Me";
  const dispatch =useDispatch();
  const handleRequestSpecLists=(length:number)=>dispatch({type:ListActionsTypes.GET_SPEC_LIST,length})
  const list = useSelector((state:AppState)=>state.listReducer.payload);
  const emitEvent = (length:number)=>handleRequestSpecLists(length);

  return (
    <AppContainer data-test="appContainer">
      <Header />
      <section className="main">
        <HeaderLine  header={header} desc={desc} tempArr={tempArr}/>
        <Button buttonText={buttonText} emitEvent={emitEvent}/>
        <ListItem list={list}/>
      </section>
    </AppContainer>
  );
}

export default memo(App);
