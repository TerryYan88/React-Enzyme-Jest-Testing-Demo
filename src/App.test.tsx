import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import renderer from 'react-test-renderer';
import App from './App';
import { AppContainer } from "./AppStyle";
import { findByTestAttr } from "./Utils";
import { Header } from "./components/header";
import { HeaderLine } from "./components/headerline";
import { HeaderLineContainer } from "./components/headerline/headerlineStyle";
import { HeaderContainer } from "./components/header/headerStyle";
import { Provider } from 'react-redux';
import { store } from './redux/rootStore';

interface ItempArr {
  fName: string,
  lName: string,
  email: string,
  age: number,
  onlineStatus: boolean,
}
interface HeaderLineProps {
  header: string;
  desc: string;
  tempArr: Array<ItempArr>;
}

const setUpHeaderLine = (props: HeaderLineProps) => {
  const component = shallow(<HeaderLine {...props} />)
  return component;
}

const setUp = (props = { }) => {
  const component = shallow(
    <Provider store={store}>
      <App {...props} />
    </Provider>
  )
  return component;
}
describe('renders app', () => {
  it("should show a app", () => {
    const component = (
      <Provider store={store}>
        <App />
      </Provider>
    );
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  })
  describe('app container', () => {
    let component: ShallowWrapper;
    beforeEach(() => {
      component = setUp();
    })
    it("should app to render without errors", () => {
      const app = component.find(App);
      expect(app.length).toBe(1);
      // console.log(pp.length)
      // const app = component.find(AppContainer);
      // expect(app.length).toBe(1);
      // const appContainer = findByTestAttr(component, "appContainer")
      // expect(appContainer.length).toBe(1);
    })
  })


  describe("header component", () => {
    const setUp = (props = { }) => {
      const component = shallow(<Header {...props} />) // can test props 
      return component;
    }
    let component: ShallowWrapper;
    //beforeEach(): before every single test 
    //all declarations should pass function  const component = shallow(<Header {...props}/>) first
    beforeEach(() => {
      component = setUp();
    })

    it("should render without errors", () => {
      const container = component.find(HeaderContainer);
      expect(container.length).toBe(1);
      const wrapper = findByTestAttr(component, "headerComponent");
      expect(wrapper.length).toBe(1);
    })
    it("should render a logo", () => {
      //data-test 
      const logo = findByTestAttr(component, "logoImg");
      expect(logo.length).toBe(1);
    })
  })




  describe("headerline component", () => {
    let component: ShallowWrapper;
    beforeEach(() => {
      const props: HeaderLineProps = {
        header: "Test Header",
        desc: "Test Dest",
        tempArr: [{
          fName: "Test fName",
          lName: "Test lName",
          email: "Test email",
          age: 0,
          onlineStatus: true,
        }]
      }
      component = setUpHeaderLine(props);
    })
    it("should render without errors", () => {
      const container = component.find(HeaderLineContainer);
      expect(container.length).toBe(1);
      const headLineComponent = findByTestAttr(component, "HeadlineComponent");
      expect(headLineComponent.length).toBe(1);
    })
    it("should render a H1", () => {
      const h1 = findByTestAttr(component, "header");
      expect(h1.length).toBe(1);
    })
    it("should render a desc", () => {
      const desc = findByTestAttr(component, "desc");
      expect(desc.length).toBe(1);
    })
  })
});
