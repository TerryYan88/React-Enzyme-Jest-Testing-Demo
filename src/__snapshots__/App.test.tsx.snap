// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders app should show a app 1`] = `
.c1 {
  display: block;
  width: 100%;
  height: 50px;
  margin: 0 auto;
  padding: 0 10px;
  background-color: black;
}

.c1 .wrap {
  position: relative;
  display: block;
  width: 100%;
  height: 100%;
  margin: 0 auto;
  max-width: 980px;
}

.c1 .wrap .logo {
  display: block;
  width: 250px;
  position: absolute;
  left: 10px;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
}

.c1 .wrap .logo img {
  width: 20%;
  margin: 0;
}

.c2 {
  border: none;
  width: 200px;
  padding: 12px 6px;
  background-color: pink;
  font-size: 24px;
  font-weight: 500;
  border-radius: 10px;
}

.c0 .main {
  width: 100%;
  max-width: 980px;
  margin: 50px auto 80px;
}

<div
  className="c0"
  data-test="appContainer"
>
  <div
    className="c1"
    data-test="headerComponent"
  >
    <div
      className="wrap"
    >
      <div
        className="logo"
      >
        <img
          alt="Logo"
          data-test="logoImg"
          src="http://localhost:3000/assets/2.png"
        />
      </div>
    </div>
  </div>
  <section
    className="main"
  >
    <div
      className=""
      data-test="HeadlineComponent"
    >
      <h1
        data-test="header"
      >
        Posts
      </h1>
      <p
        data-test="desc"
      >
        Click the button to render posts!
      </p>
    </div>
    <button
      className="c2"
      data-test="buttonComponent"
      onClick={[Function]}
    >
      Click Me
    </button>
  </section>
</div>
`;
